src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -105,30 +105,28 @@ void ZooKeeper::init(ZooKeeperArgs args_)
                 throw KeeperException("Cannot use any of provided ZooKeeper nodes", Coordination::Error::ZCONNECTIONLOSS);
         }
 
-        impl = std::make_unique<Coordination::ZooKeeper>(nodes, args, zk_log);
+        impl = std::make_unique<Coordination::ZooKeeper>(nodes, args, zk_log, [this](size_t node_idx, const Coordination::ZooKeeper::Node & node)
+            {
+                connected_zk_host = node.address.host().toString();
+                connected_zk_port = node.address.port();
+
+                connected_zk_index = node_idx;
+            });
 
         if (args.chroot.empty())
             LOG_TRACE(log, "Initialized, hosts: {}", fmt::join(args.hosts, ","));
         else
             LOG_TRACE(log, "Initialized, hosts: {}, chroot: {}", fmt::join(args.hosts, ","), args.chroot);
 
-        Poco::Net::SocketAddress address = impl->getConnectedAddress();
-
-        connected_zk_host = address.host().toString();
-        connected_zk_port = address.port();
-
-        connected_zk_index = 0;
-
-        if (args.hosts.size() > 1)
+        if (connected_zk_index != 0)
         {
-            for (size_t i = 0; i < args.hosts.size(); i++)
+            std::uniform_int_distribution<UInt32> fallback_session_lifetime_distribution
             {
-                if (args.hosts[i] == address.toString())
```

Comment:
This did not work for cases where ZooKeeper host is a DNS name (including integration tests)